falco talon v0.1.0 (#1363)
falcosecurity/falco-website: content/en/blog/falco-talon-v0-1-0/index.md

All these methods rely on a 3rd party FaaS (*Function as a Service*) and come with drawbacks, all actions must be developped by the users to manage:

* The errors
* The Falco event format
* The authentication
* The K8s SDK complexity
* The security
* The upgrades of the dependencies
* Latency
* Complexity to manage sequential actions
* Intrication between the function and the configuration

**This is why we started to develop a custom solution specifically built for Falco: `Falco Talon`**:

* Tailor made for the Falco events
* Easy to define rules
* No-code implementation for end-users
* UX close to Falco with the rules (yaml files with append, override mechanisms)
* Allow to set up sequential actions to run
* Structured logs (with a trace id)
* OTEL/Prometheus Metrics and OTEL Traces

### What is it good for?

* React in real-time to the Falco Events
* Allow fine granularity to match the events to react to
* Responding to default rules with specific overrides

### What is it not (yet?) good for?

* Complex reaction worflows with conditions between the steps

> * Act outside a Kubernetes context (no ssh-through actions like Ansible)

Reviewer: What does it mean? Does it mean that I cannot install this yet on a node without k8s? In that case it should be better specified for people that do not know what ssh-through actions are.

## Docs

> A dedicated website has been created to host the documentation, you find it at this address: [https://docs.falco-talon.org](https://docs.falco-talon.org).

Reviewer: ```suggestion
A dedicated website has been created to host the documentation: [https://docs.falco-talon.org](https://docs.falco-talon.org).
```

## How Falco Talon works

> As the same manner Falcosidekick works, `Falco Talon` receives the events from Falco by http. All you have to do, to connect Falco and `Falco Talon`, is to set in your `falco.yaml`:

Reviewer: ```suggestion
As the same manner Falcosidekick works, `Falco Talon` receives the events from Falco by http. All you have to do to connect Falco and `Falco Talon` is to set in your `falco.yaml`:
```

```yaml
jsonOutput: true
jsonIncludeOutputProperty: true
httpOutput:
  enabled: true
  url: "http://<falco-talon>:2803/"
```

If you already use Falcosidekick to forward your Falco events to the world, an integration is available since [Falcosidekick 2.29.0](https://github.com/falcosecurity/falcosidekick/releases/tag/2.29.0):

```yaml
talon:
  address: "http://<falco-talon>:2803/"
  checkcert: false
```

When the events are received by `Falco Talon`, an internal queue system based on NATS Jetstream is in charge to deduplicate them, to avoid to trigger the same action for the same cause for nothing.

`Falco Talon` will then compare the event with the rules created by the user, if an event matches with a rule, a series of actions are sequentially performed. At the end of each step, a notification with the status is sent, and a log is emmited.

### Rules

> The rules are the "core" of `Falco Talon`, the way it has to know which actions to trigger for which Falco event.

Reviewer: ```suggestion
The rules are the "core" of `Falco Talon` as they describe which actions to trigger for which Falco event.
```

> All rules are written as yaml file, evaluated in the order they are given to `Falco Talon` (as arguments or in the config file). Te latest rules overriding the previous ones.

Reviewer: ```suggestion
All rules are written as yaml file, evaluated in the order they are given to `Falco Talon` (as arguments or in the config file), with rules specified later in the file overriding the previous ones.
```

The rules are composed of 2 blocks:
* the `action` block defines which **actionner** to use with its parameters, this block can be imported by multiple rules (like the `macros` can be used in the Falco rules)
* the `rule` block defines the criterias to match to trigger the actions

The criterias to match the event with the actions can use all elements that compose a Falco event JSON payload:
* the Falco rule name
* the `priority`
* the `tags`
* the `output fields`

#### Examples

{{% pageinfo color="info" %}}

> When `Falco Talon` receives an event triggered by the Falco rule named `Terminal shell in container`, and this event doesn't concernd the kubernetes namespaces `kube-system` and `falco`, then the related pod is labeled `suspicious: true`

Reviewer: ```suggestion
When `Falco Talon` receives an event triggered by the Falco rule named `Terminal shell in container`, and this event doesn't concern the kubernetes namespaces `kube-system` and `falco`, then the related pod is labeled `suspicious: true`
```

### Outputs

Some `actionners` require an `output`, an `output` is a target for the artifact created by the `actionner`, for example for the file retrieved by `kubernetes:download` or the `.pcap` created by `kubernetes:tcpdump`.

3 `outputs` are available today:
- `local:file` (only useful for local tests)
- `aws:s3`
- `minio:s3`

The list of the available `outputs` can be found on [docs/outputs](https://docs.falco-talon.org/docs/outputs/list/).

#### Example

```yaml
- rule: Redirect STDOUT/STDIN to Network Connection in Container
  match:
    rules:
      - Redirect STDOUT/STDIN to Network Connection in Container
  actions:
    - action: Run tcpdump for 5s
      actionner: "kubernetes:tcpdump"
      parameters:
        snaplen: 512
        duration: 5
      output:
        target: "aws:s3"
        parameters:
          bucket: <my-bucket>
          prefix: /tcpdump/
          region: us-east-1
```

![aws s3 with .pcap](/blog/falco-talon-v0-1-0/images/awss3.png)

### Notifiers

Even we're talking about a "response engine", a framework to automatically react to some events, we still want (we humans), to be noticed of what's happening or keep traces of the performed actions.

> Moreover the logs emitted in the *stdoud*, some `notifiers` can be used to forward the resultsof the actions:

Reviewer: ```suggestion
Apart from logs output to `stdout`, some `notifiers` can be used to forward action results:
```

#### Examples

`k8sevents`:

```yaml
action: kubernetes:tcpdumpthought, 
apiVersion: v1
eventTime: "2024-09-05T12:52:10.819462Z"
firstTimestamp: null
involvedObject:
  kind: Pod
  namespace: default
kind: Event
lastTimestamp: null
message: |
  Status: success
  Message: action
  Rule: Redirect STDOUT/STDIN to Network Connection in Container
  Action: Run tcpdump for 5s
  Actionner: kubernetes:tcpdump
  Event: Redirect STDOUT/STDIN to Network Connection in Container
  Namespace: default
  Pod: cncf-55696bc998-5xjcb
  Output: a tcpdump "tcpdump.pcap" has been created
  TraceID: c954bd8b3391a08f23079552fdd639f3
metadata:
  creationTimestamp: "2024-09-05T12:52:10Z"
  generateName: falco-talon-
  name: falco-talon-zgxfm
  namespace: default
  resourceVersion: "115862544"
  uid: 3b4bd17f-ed1a-4693-bfd7-d10f674a8008
reason: falco-talon:action:kubernetes:tcpdump:success
reportingComponent: falcosecurity.org/falco-talon
reportingInstance: falco-talon
source:
  component: falco-talon
type: Normal
```

`slack`:

![slack](/blog/falco-talon-v0-1-0/images/slack.png)

## A tool designed for the production

> I spent 10 years of my career as a DevOps/SRE, managing traditional and cloud infrastructures, I know what pain it is to manage systems not well designed for the runtime. This is why we tried from the beginning to create a tool easy to rule all along it lifecycle.

Reviewer: ```suggestion
I spent 10 years of my career as a DevOps/SRE, managing traditional and cloud infrastructures, I know how painful it is to manage systems not well designed for the runtime. This is why we tried from the beginning to create a tool easy to rule all along it lifecycle.
```

2024-09-05T14:52:10+02:00 INF notification action="Run tcpdump for 5s" actionner=kubernetes:tcpdump notifier=k8sevents rule="Reverse shell detected" stage=action status=success trace_id=c954bd8b3391a08f23079552fdd639f3

2024-09-05T14:52:11+02:00 INF output action="Run tcpdump for 5s" bucket=xxxxx category=aws file=tcpdump.pcap key=2024-09-05T14-52-10Z_default_cncf-55696bc998-5xjcb_tcpdump.pcap output="the file 'tcpdump.pcap' has been uploaded as the key 'tcpdump/2024-09-05T14-52-10Z_default_cncf-55696bc998-5xjcb_tcpdump.pcap' to the bucket 'xxxxx'" output_target=aws:s3 prefix=tcpdump/ region=us-east-1 status=success trace_id=c954bd8b3391a08f23079552fdd639f3

2024-09-05T14:52:11+02:00 INF notification action="Run tcpdump for 5s" actionner=kubernetes:tcpdump notifier=k8sevents output_target=aws:s3 rule="Reverse shell detected" stage=output status=success trace_id=c954bd8b3391a08f23079552fdd639f3
```

### Metrics

> `Falco Talon` exposes the traditionnal `/metrics` endpoint with metrics in the Prometheus format.

Reviewer: ```suggestion
`Falco Talon` exposes the traditional `/metrics` endpoint with metrics in the Prometheus format.
```
